### PR TITLE
add slash commands

### PR DESCRIPTION
**Pull request by suneettipirneni**
Plugins that implement `ISlashCommand` will support slash commands.

### CODE REVIEW

File: .env.sample [hunk 1]
```diff
@@ -12,5 +12,6 @@ WARNINGS_THRESH=x
 PAPERTRAIL_HOST=x
 PAPERTRAIL_PORT=x
 TWITTER_BEARER_TOKEN=x
+GUILD_ID=852656268090802263=x
```
Comment (c650) [inline .env.sample:15]: Nix

File: src/app/handlers/react.handler.ts [hunk 2]
```diff
@@ -42,7 +49,7 @@ export class ReactHandler extends Handler {
     await this.container.warningService.deleteChan(user.id);
   }
 
-  private _handleClassChannelPinRequest(message: IMessage, channel: TextChannel) {
+  private _handleClassChannelPinRequest(message: IMessage | PartialMessage, channel: TextChannel) {
```
Comment (c650) [inline src/app/handlers/react.handler.ts:52]: Why do you need this

File: src/app/jobs/poll.job.ts [hunk 3]
```diff
@@ -21,7 +21,7 @@ export class PollJob extends Job {
         .map(async (poll) => {
           const embed = container.pollService.createResultEmbed(poll);
 
-          await poll.msg.channel.send({embeds:[embed]}).then(() => {
```
Comment (c650) [inline src/app/jobs/poll.job.ts:24]: Nix

File: src/app/listener.ts [hunk 4]
```diff
@@ -42,6 +45,38 @@ export class Listener {
...
+        })
+        .map((entry) => {
+          const [name, plugin] = entry;
+          const options = isSlashCommand(plugin)
+            ? (plugin as unknown as ISlashPlugin).parameters
+            : undefined;
+
+          return {
+            name: name,
+            description: plugin.description.substring(0, 99),
```
Comment (c650) [inline src/app/listener.ts:61]: 99?

File: .env.sample (hunk 1 above)
Comment (suneettipirneni) [reply .env.sample:15]: This is used for local development of slash commands. Since slash commands need to be deployed, as developer will enter the guild ID they are developing in and lion will deploy commands in said guild so they can test them,

File: src/app/listener.ts [hunk 5]
```diff
@@ -42,6 +45,38 @@ export class Listener {
...
+      // We only want to use guild commands for development, because global app commands
+      // can take over an hour to propogate.
+      if (process.env.NODE_ENV === Mode.Development) {
+        if (!process.env.GUILD_ID) {
+          throw new Error('You need to set the GUILD_ID in your .env file!');
+        }
+
+        await this.container.clientService.guilds.cache
+          .get(process.env.GUILD_ID)
+          ?.commands.set(commands);
```
Comment (c650) [inline src/app/listener.ts:75]: Why do the docs have all this REST garbage If you can do this natively

File: src/app/listener.ts (hunk 4 above)
Comment (suneettipirneni) [reply src/app/listener.ts:61]: Slash commands have a description string length limit of 99 chars.

File: src/app/plugins/broadcast.plugin.ts [hunk 6]
```diff
@@ -17,7 +23,7 @@ export default class BroadcastPlugin extends Plugin {
   public override minRoleToRun: RoleType = RoleType.Admin;
   public override commandPattern: RegExp = /((message|classes)\s.+|attach|confirm|cancel)/;
 
-  private _CHANS_TO_SEND: GuildChannel[] = [];
+  private _CHANS_TO_SEND: (GuildChannel | ThreadChannel)[] = [];
```
Comment (c650) [inline src/app/plugins/broadcast.plugin.ts:26]: Y

File: src/common/types.ts [hunk 7]
```diff
@@ -95,7 +96,7 @@ export interface IPlugin extends IRunnable {
   usableInDM?: boolean;
   usableInGuild?: boolean;
   validate(message: IMessage, args: string[]): boolean;
-  hasPermission(message: IMessage): true | string;
+  hasPermission(message: IMessage | discord.CommandInteraction): true | string;
```
Comment (c650) [inline src/common/types.ts:99]: I thought you could put permissions in the configuration for slash commands? Then we don’t need to do our own validation ?

File: src/services/guild.service.ts [hunk 8]
```diff
@@ -9,7 +9,7 @@ export class GuildService {
     [Constants.Roles.Unverifed]: undefined,
   };
 
-  private _channelCache: Record<string, Maybe<GuildChannel>> = {
+  private _channelCache: Record<string, Maybe<GuildChannel | ThreadChannel>> = {
```
Comment (c650) [inline src/services/guild.service.ts:12]: Seems unrelated? Or is it because of the recent change … ugh. Should be a separate PR

**Review by c650 — CHANGES_REQUESTED** [review]

File: src/app/listener.ts (hunk 5 above)
Comment (suneettipirneni) [reply src/app/listener.ts:75]: ¯\_(ツ)_/¯

File: src/common/types.ts (hunk 7 above)
Comment (suneettipirneni) [reply src/common/types.ts:99]: Those validations are client-side validation checking the format. In our case we'd need to use lion's system for validation since it provides custom messages in chat for permissions errors.

File: .env.sample (hunk 1 above)
Comment (c650) [reply .env.sample:15]: why not loop over the guilds like i do in https://github.com/cs-discord-at-ucf/lion/pull/738/files#diff-13c50b4cc67456d5ba9f4608265a1f770f856e4ab28225bf35ba2e4bfec7bea6

File: src/common/types.ts [hunk 9]
```diff
@@ -223,6 +224,21 @@ export type ServerInfoType = 'MemberCount';
...
 export type Maybe<T> = T | undefined | null;
 
+export type MessageSendData =
+  | string
+  | discord.MessagePayload
+  | (discord.ReplyMessageOptions & { split?: false }); // Discord v13 change
+
+export type MessageEditData = string | discord.MessagePayload;
+
+export function isSlashCommand(plugin: unknown): boolean {
```
Comment (c650) [inline src/common/types.ts:234]: does this function belong in types.ts ?

File: src/app/handlers/command.handler.ts [hunk 10]
```diff
@@ -14,30 +30,40 @@ export class CommandHandler extends Handler {
     super();
   }
 
-  public async execute(message: types.IMessage): Promise<void> {
-    const command = this.build(message.content);
+  public async execute(message: types.IMessage | CommandInteraction): Promise<void> {
```
Comment (c650) [inline src/app/handlers/command.handler.ts:33]: instead of a union type, we should (upstream) populate some CommandPayloadCommon so that it has the same data whether it's from a '!' message or a slash command invocation

File: .env.sample (hunk 1 above)
Comment (c650) [inline .env.sample:15]: i prefer that we automatically get the guild from the guild service on startup, de-register all commands, and then re-register whichever ones still exist.

**Review by c650 — CHANGES_REQUESTED** [review]

**Comment by c650** [comment]
just wondering where we wanna go with this?

File: src/app/handlers/command.handler.ts (hunk 10 above)
Comment (suneettipirneni) [reply src/app/handlers/command.handler.ts:33]: I was thinking about this, and it feels like that would actually introduce more complexity. The reason being is that typescript unions automatically allow you access common properties. Also this is the only function where there's unioned command types. After this point they're separated and narrowed to their specific types. So another abstracted class would essentially be doing the same de-narrowing this function does but it would only be used in a single place.

File: src/app/handlers/command.handler.ts [hunk 11]
```diff
@@ -1,8 +1,24 @@
...
+  | {
+      type: 'message';
+      message: types.IMessage;
+      plugin: types.IPlugin;
+      command: types.ICommand;
+      isDM: boolean;
+    }
+  | {
+      type: 'interaction';
+      message: CommandInteraction;
```
Comment (c650) [inline src/app/handlers/command.handler.ts:18]: do we want to have a `types.ICommandInteraction`?

File: src/app/handlers/command.handler.ts [hunk 12]
```diff
@@ -14,30 +30,40 @@ export class CommandHandler extends Handler {
...
-    // checks to see if the user is actually talking to the bot
-    if (!command) {
-      return;
-    }
-
-    if (message.mentions?.everyone) {
-      message.reply('You cannot use a plugin, while pinging everyone.');
-      return;
-    }
+    let plugin;
```
Comment (c650) [inline src/app/handlers/command.handler.ts:37]: do you need to declare this here? it seems like you can tightly scope it to each branch of the if/else

File: src/app/handlers/command.handler.ts [hunk 13]
```diff
@@ -114,25 +146,32 @@ export class CommandHandler extends Handler {
...
+        return;
+      }
+    }
+
+    let user;
+
+    if (message instanceof CommandInteraction) {
+      user = message.user.tag;
+    } else {
+      user = message.author.tag;
```
Comment (c650) [inline src/app/handlers/command.handler.ts:174]: can we just make this a const ... = ternary

File: src/app/handlers/command.handler.ts [hunk 14]
```diff
@@ -143,13 +182,26 @@ export class CommandHandler extends Handler {
...
+      // If it's a regular command, use the regular executor.
+      if (commandData.type === 'message') {
+        await plugin.execute(commandData.message, commandData.command.args);
+      }
+
+      // If it's a slash command, use the slash runner.
+
+      // This should never happen.
+      if (!types.isSlashCommand(plugin)) {
+        throw new Error('Cannot invoke slash command on a non-slash-command plugin.');
```
Comment (c650) [inline src/app/handlers/command.handler.ts:201]: let's just log and return instead of throwing

File: src/app/handlers/command.handler.ts [hunk 15]
```diff
@@ -143,13 +182,26 @@ export class CommandHandler extends Handler {
...
+      }
+
+      await (plugin as unknown as ISlashPlugin).run(message as CommandInteraction);
```
Comment (c650) [inline src/app/handlers/command.handler.ts:204]: ?

File: src/app/index.d.ts [hunk 16]
```diff
@@ -23,6 +25,7 @@ declare global {
       TWITTER_BEARER_TOKEN?: string;
       WOLFRAM_APPID?: string;
       TESTER_TOKEN?: string;
+      GUILD_ID?: Snowflake;
```
Comment (c650) [inline src/app/index.d.ts:28]: you shouldnt need these changes anymore.

File: src/app/listener.ts [hunk 17]
```diff
@@ -42,6 +45,29 @@ export class Listener {
...
 
+      const commands = Object.entries(this.container.pluginService.plugins)
+        .filter((entry) => {
+          const [, plugin] = entry;
+          return isSlashCommand(plugin);
+        })
+        .map((entry) => {
+          const [name, plugin] = entry;
+          const options = isSlashCommand(plugin)
+            ? (plugin as unknown as ISlashPlugin).parameters
```
Comment (c650) [inline src/app/listener.ts:56]: since you filter, do you need to do this?

File: src/app/listener.ts [hunk 18]
```diff
@@ -42,6 +45,29 @@ export class Listener {
...
+            : undefined;
```
Comment (c650) [inline src/app/listener.ts:57]: use null instead for user-specified "undef"

File: src/app/listener.ts [hunk 19]
```diff
@@ -42,6 +45,29 @@ export class Listener {
...
+          return {
+            name: name,
+            description: plugin.description.substring(0, 99),
+            options,
+          };
+        });
+
+      // Register commands for all guilds.
+      for (const guild of this.container.clientService.guilds.cache.values()) {
+        guild.commands.set(commands);
```
Comment (c650) [inline src/app/listener.ts:68]: is this synchronous ? 

also do we have to do a pass of unregistering? otherwise a command that gets removed could be left dangling.

File: src/common/plugin.ts [hunk 20]
```diff
@@ -84,12 +85,18 @@ export abstract class Plugin implements IPlugin {
     const totalChannels = channels.length;
     channels.splice(this._numChannelsShown);
 
+    if (message instanceof CommandInteraction) {
+      this.container.guildService.getChannel('test')
```
Comment (c650) [inline src/common/plugin.ts:89]: ?

File: src/common/types.ts [hunk 21]
```diff
@@ -223,6 +224,21 @@ export type ServerInfoType = 'MemberCount';
 export type RoleTypeKey = keyof typeof RoleType;
 export type Maybe<T> = T | undefined | null;
 
+export type MessageSendData =
+  | string
+  | discord.MessagePayload
+  | (discord.ReplyMessageOptions & { split?: false }); // Discord v13 change
```
Comment (c650) [inline src/common/types.ts:230]: can this go in a smaller change since it's a v13 thing?

File: tsconfig.json [hunk 22]
```diff
@@ -10,6 +10,7 @@
     "types": ["node", "jest"],
     "esModuleInterop": true,
     "noImplicitOverride": true,
+    "skipLibCheck": true,
```
Comment (c650) [inline tsconfig.json:13]: what's this do ?

File: src/app/listener.ts (hunk 17 above)
Comment (tanndlin) [reply src/app/listener.ts:56]: Don't think TS updates types of an array after using a filter

File: src/app/listener.ts (hunk 19 above)
Comment (tanndlin) [reply src/app/listener.ts:68]: Set should be a strict equals to, not add.
So any slash command that was not in the set will be removed

**Comment by tanndlin** [comment]
Would it make more sense to have 2 separate handlers, one for slash one for legacy? Instead of binding them both to the same class and having to type check everywhere

**Comment by joeydotdev** [comment]
> Would it make more sense to have 2 separate handlers, one for slash one for legacy? Instead of binding them both to the same class and having to type check everywhere

+1 to this. Especially if we intend to do any enhancements on either of the types of commands, it's easier way to introduce regressions if the code is tightly coupled with each other. I think it's fine to duplicate code in this scenario if it's required

File: src/app/listener.ts (hunk 17 above)
Comment (c650) [reply src/app/listener.ts:56]: cleaner to do a map after the filter then, if filter_map doesnt exist in ts

File: tsconfig.json (hunk 22 above)
Comment (rob-3) [reply tsconfig.json:13]: https://www.typescriptlang.org/tsconfig#skipLibCheck

File: src/common/plugin.ts (hunk 20 above)
Comment (rob-3) [reply src/common/plugin.ts:89]: I'm removing this in my branch; I think it was just for testing

File: src/app/handlers/command.handler.ts (hunk 11 above)
Comment (rob-3) [reply src/app/handlers/command.handler.ts:18]: What would be the idea here? Just a wrapper around the `CommandInteraction` type? If so, seems like something we can add later if we need it.

File: src/app/handlers/command.handler.ts (hunk 12 above)
Comment (rob-3) [reply src/app/handlers/command.handler.ts:37]: Doen

File: src/app/handlers/command.handler.ts (hunk 15 above)
Comment (rob-3) [reply src/app/handlers/command.handler.ts:204]: There's some trickiness going on with how exactly we want to tag slash command plugins vs classic plugins. I can rework some of this if you don't like how it is currently.

File: src/app/listener.ts (hunk 18 above)
Comment (rob-3) [reply src/app/listener.ts:57]: This breaks the typings on line 68

File: src/app/listener.ts (hunk 19 above)
Comment (rob-3) [reply src/app/listener.ts:68]: I'm adding await just to be sure here, +1 on tanndlin's comment

File: src/common/types.ts (hunk 21 above)
Comment (rob-3) [reply src/common/types.ts:230]: Sure
